### ISSUE TITLE
Create validation of models

### ISSUE THREAD

**Issue by istvanp1** [skpref/skpref#11]

**Comment by istvanp1** [comment]
Dependent on #36

**Comment by istvanp1** [comment]
On Second thought this might not be as simple an idea as it initially seemed. What we would like to ideally have is a function that makes it easy to compare the accuracy of different models. An ideal output would be something like this:
![image](https://user-images.githubusercontent.com/25740842/96106736-72b0a680-0ed3-11eb-9686-38d3c5ec5a00.png)

To be able to do this a function would need the following inputs:
The type of task on which accuracy is being assessed, the metric based on which we are assessing the accuracy, the predictions and the ground truth.

#### Option 1
Most of this information is stored in the task. Except for the predictions. I'm wondering if there should be an option to store predictions in the task for each predict_task function.
Then the predict function would need to change to something like this:
```python
def predict_task(self, task: skpref.PrefTask, pred_name: str = None) -> numpy.array:
    # after some actual that creates task packed predictions called packed_preds
    if pred_name is not None:
        task.predictions{pred_name} = packed_preds
    return packed_preds
```
The evaluation code would look like this
```python
def evaluate(metric: callable, task: skpref.PrefTask, predictions: List[str] = 'all') -> pandas.DataFrame:
    """
    metric: is a function that calculates the accuracy probably from scikit-learn.metrics
    task: the task the analysis has been done with
    predictions: strings under which predictions have been named
    """
    if predictions == "all":
        predictions = task.predictions.keys()
    for pred in predictions:
        metric(task.predictions[pred], task.primary_table[primary_table_target_name])
  # etc
```

#### Option 2
The evaluation code takes everything a bit more explicitly
```python
def evaluate(metric: callable, task: skpref.PrefTaskType, df: pandas.DataFrame, predictions: List[str],
    ground_truth = 'str') -> pandas.DataFrame:
    """
    metric: is a function that calculates the accuracy probably from scikit-learn.metrics
    task: type of task that predictions have been made for
    df: DataFrame that contains ground truth and predictions
    predicitons: List of the column names that contain predictions
    ground_truth: column name that contains the ground truth
    """

```
@fkiraly what do you think? I think I prefer option 1 feels a bit cleaner and more in sync with the whole task based architecture, but then our evaluation function is also going to require people to use the task infrastructure, which I think is fine, the whole package is being built around it anyways.

**Comment by fkiraly** [comment]
Let´s talk about this in our meeting today.

Quick question: why does option 1 `predict_task` return a `np.array`? Aren't predictions in the general case rankings or preferences? I thought the whole idea of tasks etc was meant to be able to encode these explicitly instead of implicitly by a coding convention in arrays.

**Comment by fkiraly** [comment]
From our discussion:
* I think we first need to be clear about what `predict` functions return - are these `numpy.array`, or are we using `PosetVector` or similar? Might be worth writing up a quick specification for the class template here, explicit with input/return types
* depending on what `predict` functions return, this will imply a `metric` type signature, and there should be a small design on how/what to implement here, and how relates to `sklearn` `metrics`, e.g., wrapping them to obtain metrics for choices/preferences?

**Comment by mloning** [comment]
I'd separate results from tasks. You could have a results wrapper which collects  the data and meta data for the column names that contain predictions and the column name that contains the ground truth.

**Comment by fkiraly** [comment]
I agree with @mloning - it may have been a design flaw to not separate the "role of data" from the "data including schema and relational struture".
Not sure what the best form of this would be though.

**Comment by fkiraly** [comment]
(@mloning, are you secretly watching this??)
